### PR TITLE
[TextServer] Implement ICU/UAX#31 based `is_valid_identifier` function.

### PR DESCRIPTION
**Pull request by bruvzg**
Partially implements https://github.com/godotengine/godot-proposals/issues/916

Implements `TextServer::is_valid_identifier` function, but do not change any of the GDScript related code.

TextServerAdvnaced version use following rules - a valid identifier must:
  - Conform to normalization form C.
  - Begin with a Unicode character of class XID_Start or `"_"`.
  - May contain Unicode characters of class XID_Continue in the other positions.
  - Use UAX#31 recommended scripts only (mixed scripts are allowed).

Fallback version uses `String::is_valid_identifier` which implements following rules - a valid identifier must:
  - Begin with a Basic Latin letter character or `"_"`.
  - May contain Basic Latin letter characters, digits or `"_"` in the other positions.

### CODE REVIEW

File: doc/classes/TextServer.xml [hunk 1]
```diff
@@ -798,6 +798,21 @@
 				Returns [code]true[/code] if locale is right-to-left.
 			</description>
 		</method>
+		<method name="is_valid_identifier" qualifiers="const">
+			<return type="bool" />
+			<argument index="0" name="string" type="String" />
+			<description>
+				Returns [code]true[/code] is [code]string[/code] is a valid identifier.
+				If text server supports [code]FEATURE_BIDI_LAYOUT[/code] feature, a valid identifier must:
```
Comment (akien-mga) [inline doc/classes/TextServer.xml:806]:
```suggestion
				If the text server supports the [constant FEATURE_BIDI_LAYOUT] feature, a valid identifier must:
```

File: doc/classes/TextServer.xml [hunk 2]
```diff
@@ -798,6 +798,21 @@
...
+				- Conform to normalization form C.
+				- Begin with a Unicode character of class XID_Start or [code]"_"[/code].
+				- May contain Unicode characters of class XID_Continue in the other positions.
+				- Use UAX #31 recommended scripts only (mixed scripts are allowed).
+				If [code]FEATURE_BIDI_LAYOUT[/code] feature is not supported, a valid identifier must:
```
Comment (akien-mga) [inline doc/classes/TextServer.xml:811]:
```suggestion
				If the [constant FEATURE_BIDI_LAYOUT] feature is not supported, a valid identifier must:
```

**Comment by timothyqiu** [comment]
IINM, `TextServer` is for "rendering" fonts and text layouts. I think putting `is_valid_identifier` here might cause some trouble. e.g. A plugin using Unicode identifiers internally might not work in projects using an incompatible text driver.

A valid GDScript should be always valid no matter using advanced text server or not.

**Comment by reduz** [comment]
Is this missing anything?

**Review by reduz — APPROVED** [review]

**Comment by bruvzg** [comment]
> Is this missing anything?

No, it's ready, but it and #54905 are pretty useless by itself, until Unicode support is implemented in the GDScript parser (not sure what is the final decision about it). Also, since it a part of optional TextServer, using it would make GDScript parser dependent on a specific TextServer implementation.

**Comment by reduz** [comment]
Code looks good to me, but left to assess that this approach is desired.

**Comment by reduz** [comment]
@bruvzg I think we discussed at some point that for text servers that are not advanced, we could have a more hacky implementation, like using a table of valid/invalid characters that may have false positives but that since this is supposed to be checked properly in the editor, it should still be allowed.

**Comment by bruvzg** [comment]
> we could have a more hacky implementation, like using a table of valid/invalid characters that may have false positives but that since this is supposed to be checked properly in the editor, it should still be allowed.

I guess we can have `XID_Start` and `XID_Continue` checks by using simple lookup tables, and the rest (script check and normalization) is not that important. And spoofing checks (https://github.com/godotengine/godot/pull/54905) are completly optional, it's probably enough to document that it's not supported with fallback text server.

**Comment by reduz** [comment]
I think the logic we should implement is the following for GDScript, Expression and VariantParser

1) Have a function that lets you know that some character is a valid identifier character, this includes control or whathever characters, we can a big hardcoded table with ranges for this that you can generate from ICU and this is harcoded in Godot, outside TextServer.
2) For the text servers that don't have ICU or text data needed to actually determine what is a valid identifier, have a function that uses the above table and lets you know that its valid or not (even if there are false positives). For text servers that support it, do it properly.

The reason this is needed is that GDScript, Expression, VariantParser, etc. use char by char parsing to determine identifiers. So the logic should be that it first gets the whole String and then it should determine that something is a valid identifier or not.

**Comment by bruvzg** [comment]
Added `is_unicode_identifier_start` and `is_unicode_identifier_continue` (in the `char_utils.h`) functions and updated fallback text server to use it.

`TextServerAdvnaced` version use following rules - a valid identifier must:

- Conform to normalization form C.
- Begin with a Unicode character of class XID_Start or "_".
- May contain Unicode characters of class XID_Continue in the other positions.
- May contain ZWNJ/ZWJ in the special cases, see [Layout and Format Control Characters](https://www.unicode.org/reports/tr31/#Layout_and_Format_Control_Characters).
- Use UAX#31 recommended scripts only (mixed scripts are allowed).

Fallback implements the following rules - a valid identifier must:

- Begin with a Unicode character of class XID_Start or "_".
- May contain Unicode characters of class XID_Continue in the other positions.

File: core/string/char_utils.h [hunk 3]
```diff
@@ -33,6 +33,1444 @@
...
+
+static CharRange xid_start[] = {
+	{ 0x41, 0x5a },
+	{ 0x5f, 0x5f },
+	{ 0x61, 0x7a },
+	{ 0xaa, 0xaa },
+	{ 0xb5, 0xb5 },
+	{ 0xba, 0xba },
+	{ 0xc0, 0xd6 },
+	{ 0xd8, 0xf6 },
```
Comment (akien-mga) [inline core/string/char_utils.h:49]: Maybe these could go to a `core/string/char_range.inc` to reduce the clutter in `char_utils.h` (which would be included here for the two new methods that use this data)? Having so much data in a header makes the actual logic harder to consult IMO.

File: doc/classes/TextServer.xml [hunk 4]
```diff
@@ -944,6 +944,21 @@
 				Returns [code]true[/code] if locale is right-to-left.
 			</description>
 		</method>
+		<method name="is_valid_identifier" qualifiers="const">
+			<return type="bool" />
+			<argument index="0" name="string" type="String" />
+			<description>
+				Returns [code]true[/code] is [code]string[/code] is a valid identifier.
+				If the text server supports the [constant FEATURE_UNICODE_IDENTIFIERS] feature, a valid identifier must:
+				- Conform to normalization form C.
```
Comment (akien-mga) [inline doc/classes/TextServer.xml:953]:
```suggestion
				- Conform to normalization from C.
```
? Or is "form C" a type of form?

**Review by akien-mga — APPROVED** [review]

File: doc/classes/TextServer.xml (hunk 4 above)
Comment (bruvzg) [reply doc/classes/TextServer.xml:953]: It's "form C", see https://www.unicode.org/reports/tr15/#Norm_Forms

File: core/string/char_utils.h (hunk 3 above)
Comment (bruvzg) [reply core/string/char_utils.h]: Moved ranges to a separate file.

File: core/string/char_utils.h [hunk 5]
```diff
@@ -31,7 +31,25 @@
 #ifndef CHAR_UTILS_H
 #define CHAR_UTILS_H
 
-#include "core/typedefs.h"
+#include "char_range.h"
```
Comment (akien-mga) [inline core/string/char_utils.h:34]: Not too important but we usually name such data-only includes with `.inc` extension.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**